### ISSUE TITLE
Add license identiers for JavaScript used in web builds

### ISSUE THREAD

**Issue by morganwillcock** [adventuregamestudio/ags#2164]
Presumably, if the JavaScript used to run the web build of a game is not derived from the game and is considered part of the game engine, it should be possible to add license identifers so that tools which check the licensing will allow the game to run?

Is all of the JavaScript used to run a game (`ags.js` and `my_game_files.js`, plus the scripts in `index.html`) part of AGS or is it derived from something else?

**Comment by ericoporto** [comment]
So ags engine library, including it's dependencies, builds as ags.js and ags.wasm. For all purposes there isn't dynamic loading in the wasm VM. The index.html is just what I wrote to make the game start, so it's part of ags itself too.

We use Emscripten to build things, which is part of the Chromium project, so I will go as far as support whatever they support, but any random browser they ignore, we ignore too.

**Comment by morganwillcock** [comment]
I am just suggesting to add license headers to code which is released under a license, which is why I wanted to clarify where the code originates from. The manual website does have a link that a human can click to get the un-minified version of the code, but I'm not sure whether that is a requirement for automated checks or not.

> I really don't want any new weird tool introduced in the Emscripten workflow.

There isn't a tool required in any workflow, unless it is impossible to add a comment to the code. Is that the case?

> And before making any changes you should verify you are actually making something that works in Game jolt, itch io and Newgrounds, which all run the engine in an iframe. No one is actually running these builds in a plain server.

It is the client side which is responsible for reading the license identifier, whether that be a human with their eyes or an automated tool. I don't think there is any special requirement on the server side.

> I don't think the game developers have to have this forced on them

I'm not sure why developers would see any difference. The only issue would be if the code being run for the web build was theirs and released under a different license. But I think we are just discussing "engine" code here.

**Comment by ericoporto** [comment]
I think if you want that you should go upstream, to Emscripten, and get your functionality there. There won't be any clean way to get it here as far as I can think. I did a lot of work upstream to ensure we build things with the less amount of changes as possible.

And we can also have an additional wasm and js file per thread if we are building with pthreads turned on - right now, itch io is the only website that can run ags web engine when it's built with threads, this is why it isn't the default.

**Comment by ivan-mogilko** [comment]
Is not this a question about adding license headers like we have in h/cpp files, only to js and html files?
Unless I misunderstood...

**Comment by ivan-mogilko** [comment]
Do i understand correctly, that this question is referring to the game files produced by the "Web" target in the Editor?

If i get it right, what @morganwillcock is asking is, where does the contents of these files come from, is it purely generated, or based on something in our repository.

If it comes from our repository, then won't adding a license header, as a comment, to these source files automatically bring same comment to the output game files (js, and index.html)?

**Comment by morganwillcock** [comment]
> Do i understand correctly, that this question is referring to the game files produced by the "Web" target in the Editor?
> 
> If i get it right, what @morganwillcock is asking is, where do the contents of these files come from, is it purely programmatically generated, or based on something in our repository?
> 
> If it comes from our repository, then won't adding a license header, as a comment, to these source files automatically bring same comment to the output game files (js, and index.html)?

Yes, those were my questions, to ask where the files originate to see if license headers can be added.

If `ags.js` is purely Emscripten output then what is the license for the code that is generated in that file? It seems that compiling with the option `EMIT_EMSCRIPTEN_LICENSE=1` will include Emscripten's own license header in the output, which might suggest the output is (at least partially) under their license.

**Comment by ericoporto** [comment]
wasm VM is a minimal vm with few assembly instructions and no access to the DOM, a posix system, the libc, OpenGL ES headers, OpenAL, everything is smoke and mirrors in the Emscripten project. Our dependencies, which includes at least SDL2, SDL2_Sound, miniz, alfont, freetype, Xiph Ogg Vorbis and Theora, our allegro, and probably other things, then AGS engine itself, it all gets smashed together. Then this all gets instrumented because we don't have a single loop. Then things get optimized. Things that requires to escape the VM goes to the ags.js, things that can work inside the VM goes to ags.wasm. Both have code from this different sources.

Question, why does the windows port gets away with no licenses?

**Comment by ivan-mogilko** [comment]
I am honestly very confused, what is this "situation regarding license" that has to be fixed in all ports? What "things that are never updated" that has to be got rid of?

When reading the initial @morganwillcock 's post I thought he's referring to basically adding a simple comment on top of the *.js and *.html files, because these are essentially a *source code*, unlike *compiled* files like exe or wasm. At least that was my rational explanation, because I am not aware of what "tools which check the licensing" are there and what does their work implies. Maybe I was wrong and this has a different meaning?

Perhaps I am making a mistake barging into this conversation, as I may be missing information. I'd wish there was more elaborate explanation on requirements in the ticket though.

**Comment by ericoporto** [comment]
This may be a confusion on my part, because Morgan doesn't specify on his issue, but my between the lines reading is he wants the mobile port to be compatible with this https://www.gnu.org/software/librejs/index.html

This adds a bunch of weird expectations - you can only use licenses with a matching spdx indentifier (and that is also understood by the extension) and you have to identify with the license per section of the script (as I mentioned we use a bunch of different things smashed together in the same blob). The extension doesn't care if the license you add is correct actually, if the total of js lines (in separate files or in the html itself), the js can run because the extension agrees it's"free software".

I think this is unreasonable perspective from games, and afaict none of the other engines does it. I am ok with adding a human readable "something" in the menu if desired.

 > "situation regarding license" that has to be fixed in all ports

Afaik, we would need to add mention to the libraries and copyright holders of those for some libraries, even if the distribution is binary.

An example of the ones that DOESN'T is like SDL_Sound which has a clause

> If you use this software  in a product, an acknowledgment in the product documentation would be  appreciated but is not required.

Some like FreeType will require though

> If you use it, or  only parts of it,  in a program,  you must acknowledge somewhere  in  your  documentation  that  you  have  used  the  FreeType code.

But for the documentation part, it appears a mention in your manual with be sufficient. Thus the suggestion to just add a simple link to the - it's not obvious that is the documentation. But maybe it's not needed. Still, for some reason we are shipping license files in the Linux port (I don't know why), but it's weird we only add those for the things built dynamically: https://github.com/adventuregamestudio/ags/blob/e3ad2aad411ecc6ed08bdb6b6c52880e2ab0937b/Editor/AGS.Editor/BuildTargets/BuildTargetLinux.cs#L35

But it can be I am confusing how this should work and the Windows approach of not shipping any license is correct. In Android there is a screen with hardcoded text regarding license - only visible in the player app though.

**Comment by morganwillcock** [comment]
> This may be a confusion on my part, because Morgan doesn't specify on his issue, but my between the lines reading is he wants the mobile port to be compatible with this https://www.gnu.org/software/librejs/index.html

It would be a beneficial side effect if those checks passed, but the primary aim would just to indicate the license which is used. I think the identifiers themselves are fairly common. For the record, I don't actively use LibreJS, but I have checked that JavaScript used in the manual's website passes LibreJS tests. It seems a reasonable way to measure whether code that is being distributed is clearly indicating its licensing.

> I think this is unreasonable perspective from games, and afaict none of the other engines does it.

Isn't it a benefit if this engine runs where others do not?

> I am ok with adding a human readable "something" in the menu if desired.

This would be better than indicating nothing, as long as the licensing information is correct.

If using `EMIT_EMSCRIPTEN_LICENSE=1` doesn't cover the licensing of the generated file it is probably better to provide no information unless obligated to. It might be worse to misrepresent the licensing by providing the wrong information.

> Question, why does the windows port gets away with no licenses?

That is a very interesting question, when the Linux port does. I would imagine that the first person who first packaged the files for Linux thought that providing the correct licensing information was important, whereas the Windows port just relied on the telling the game developer to do the "right thing" (https://adventuregamestudio.github.io/ags-manual/DistGame.html#licensing).

**Comment by ericoporto** [comment]
> It might be worse to misrepresent the licensing by providing the wrong information.

Again, see linux port.

**Comment by morganwillcock** [comment]
My interest is in complying the licensing requirements by making any required license information available. If the checks of an automated tool block the code from running even if the licensing identifier is present, this is a separate issue and may just be a fault with the tool.

I am not interested in discussing political views.

Given what you have stated to me privately I am happy to leave the conversation at this point. Feel free to close this issue if there is no problem.

**Comment by AlanDrake** [comment]
You don't have to embed the license in the javascript source, you can just as well dump the license files in the same folder.
Or just leave it up to the user to do "what is right".

**Comment by ericoporto** [comment]
I don't want this to be hacky, if it HAS to be done, it better be in emscripten itself. I will take a stab a PRing there later, it's a few things that have to be done.

- In embuilder, there is vague license information that is only printed for listing the available ports, instead, add a specific license property to each port, and then make embuilder put this information somewhere
- Add a way to pass to EMCC additional licensing information (so one can pass the license of what is being built)
- refactor [js_manipulation.py](https://github.com/emscripten-core/emscripten/blob/main/tools/js_manipulation.py) license related functions so it can actually take these multiple licenses, and merge them and output them in the necessary places
- make sure this gets output to every js generated by emscripten tooling
- additional html parsing so it can re-add licensing there after minifying
